File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.java
class TrieNode
{
    HashMap<Character, TrieNode> children = new HashMap<>();
    boolean isEnd;
}

class WordDictionary {

    TrieNode root;
    public WordDictionary() {
        root = new TrieNode();
    }
    
    public void addWord(String word) {
        TrieNode curr = root;
        for (char c: word.toCharArray())
        {
            curr.children.putIfAbsent(c, new TrieNode());
            curr = curr.children.get(c);
        }
        curr.isEnd = true;
    }
    
    public boolean search(String word) {
        List<TrieNode> nodes = new ArrayList<>();
        nodes.add(root);
        for (char c: word.toCharArray())
        {
            List<TrieNode> nextNodes = new ArrayList<>();
                for (TrieNode node: nodes)
                {
                    if (c == '.') 
                    {
                        nextNodes.addAll(node.children.values());
                    }
                    else 
                    {
                        if (node.children.containsKey(c)) nextNodes.add(node.children.get(c));
                    }
                }
                if (nextNodes.size() == 0) return false;
                nodes = nextNodes;
        }
        
        for (TrieNode node: nodes)
            if (node.isEnd) return true;
        
        return false;
    }
}




/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False

class WordDictionary:

    def __init__(self):
        self.root = TrieNode()    

    def addWord(self, word: str) -> None:
        curr = self.root
        for c in word:
            if not c in curr.children:
                curr.children[c] = TrieNode()
            curr = curr.children[c]
        curr.isEnd = True

    def search(self, word: str) -> bool:
        currList = [self.root]
        for c in word:
            nextList = []
            for curr in currList:
                if c == '.':
                    nextList += [v for v in curr.children.values()]
                else:
                    if c in curr.children:
                        nextList.append(curr.children[c])
            currList = nextList

        for curr in currList:
            if curr.isEnd: return True

        return False


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)